hunter-packages/ogles_gpgpu
# adopt hunter.develop changes

Transfer changes from working hunter.develop branch to hunter.  This includes: updates to clang-format formatting; shader unit tests via gtest+gauze+aglet; travis updates to support osmesa testing.  A condensed git history from the hunter.develop branch is included below.

* replace astyle w/ clang-format
  + disable default platform optimizations (leave this up to the user)
  + see .clang-format for coding style
  + Use style that closely matches original author’s style
  +  *INDENT-{OFF,ONF}* -> clang-format {off,on}
  +  add .clang-format
  + add platform specific gpu dependencies
  + shader string fixes to support visual studio
    - avoid #define inside shader strings for vs-14-2015
    - use consistent top level precision declarations (drop per delcaration precision)
    - add some missing astyle `// *INDENT-{ON,OFF}* ` guards
    - avoid median filter test for `defined(_WIN32) || defined(_WIN64)` due to reliance on internal #define
* update hunter: GIT_SUBMODULE feature, cellar feature, gauze package, etc
* use gauze and aglet for cross platform shader testing
  + gauze : for cross platform tests
  + aglet : for opengl context
  + update gtests to use gauze_main() + aglet::GLContext
  + add logic to support shader GTesting
     - host builds: desktop, iOS or Android
     - ci builds: android emulator (via gauze) or linux via osmesa (cpu)
* DisplayProc: add setter for glViewport offset
* remove default (i.e., hardcoded) gpu platform optimizations in core setup
  + user must explicitly call MemTransferFactory::tryEnablePlatformOptimizations()
  + we may want to support generic opengl calls with raw images/pixel buffers on an iOS platform, for example.
* enable windows tests; consider opengl context failure a test failure
* add missing dependencies to Config.cmake.in
* fix copyright/author for new files (header copy and paste issues)
* CMakeLists.txt: add platform specific gpu dependencies
* shader string fixes to support visual studio
  + avoid #define inside shader strings for vs-14-2015
  + use consistent top level precision declarations (drop per delcaration precision)
  + add some missing astyle `// *INDENT-{ON,OFF}* ` guards
  + avoid median filter test for `defined(_WIN32) || defined(_WIN64)` due to reliance on internal and incompatible #define
* use glfw for all but ios
* remove some unused; move gl* calls from headers to cpp files
* traivs.yml
  + update to trusty + gcc-pic-hid-sections
  + remove “—strip” for os=linux to avoid issue with autoconf project
  + fix linux==travis pip3
  + update linux easy_install3
* Appveyor.yml:
  + add explicit `image: Visual Studio 2015`
  + add MSYS_PATH for appveyor
  + simplify windows build matrix; exclude pr.* from branch builds
* restrict opencv build (used for unit tests) to core + imgproc + highgui (minimal)
* add glew.h before glfw.h for ogles_gpgpu test
* update/simplify travis; update copyright info
* hunterized unit tests (gtest)
   + add core GTest unit tests (use OpenCV for IO and datatypes)
   + add custom OpenCV CMAKE_ARGS to fix opencv os x build error
   + initial per filter coverage w/ simple setup + teardown + readout test (some value based tests added)
* update .gitignore
* fix new shader author/copyright notice
* move harris corner shader to designated shader
* add simple 3x3 hessian class; output(Ixx, Iyy, Ixy, det(H))
* add rgb2hsv and hsv2rgb shaders
* add simple gain shader
* provide two output options in hessian filter
   + det(H(R)) det(H(G)) det(H(B)) 1.0
   + Ixx(R) Ixy(R), Iyy(R), det(H(R))
* add newline to logging code
* load egl{Create,Destroy,ClientWait}SyncKHR sync operators dynamically, w/ fallback to glFinish()
* add missing libs to target_link_libraries; add missing defs to target_comile_definitions
* make nms output the peak value instead of binary value
* allow control of input and output channel for NMS processing
* add delegates for generic per filter "self reflection"
   + virtual void setPreProcessCallback(ProcDelegate &cb);
   + virtual void setPostProcessCallback(ProcDelegate &cb);
   + virtual void setPreRenderCallback(ProcDelegate &cb);
   + virtual void setPostRenderCallback(ProcDelegate &cb);
   + virtual void setPreInitCallback(ProcDelegate &cb);
   + virtual void setPostInitCallback(ProcDelegate &cb);
* add access method for shader program id
* comments, set inpuPixFormat on yuv->rgb shader initialization
* static analyzer: null assertions, order of initialization
* DiffProc (add missing destructor) : required by Xcode for shared_ptr<DiffProc> use in some cases
* FirProc: add alpha  (scale) and beta (offset) shader uniforms
* remove some custom/experimental shader source
* add missing strength/scale uniform in flow shader
* clear filters from procPasses in destructor
* optical flow destructor fix: Pr.fix.flow.destruction
* move forward declaration of ANativeWindowBuffer out of ogles_gpgpu na…
* add find_package calls for android system libs  based on oglplus cmake find_package boilerplate functions
* load egl{Create,Destroy,ClientWait}SyncKHR sync operators dynamically, w/ fallback to glFinish()
* add missing libs to target_link_libraries; add missing defs to target_comile_definitions
    > error: 'eglCreateSyncKHR' was not declared in this scope
* add MemTransferOpt::flush()
 - w/ MemTransferAndroid::flush() using:
 - eglClientWaitSyncKHR(mEGLDisplay, sync, EGL_SYNC_FLUSH_COMMANDS_BIT_KHR, us);
 - instead of `glFlush()` which is prefered w/ some drivers.
* make nms output the peak value instead of binary value
* allow control of input and output channel for NMS processing
* add access method for shader program id
* comments, set inpuPixFormat on yuv->rgb shader initialization
